{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/cluster/ggs_lab/mtparker/.conda/envs/snakemake6/lib/python3.10/site-packages', '/cluster/ggs_lab/mtparker/papers/fiona/global_5ss_analysis/rules/notebook_templates']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\x85\\t\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94\\x8cA../annotations/Araport11_GFF3_genes_transposons.201606.no_chr.gtf\\x94\\x8c#annotations/athaliana_A5_strict.gtf\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x05fasta\\x94K\\x00N\\x86\\x94\\x8c\\x03gtf\\x94K\\x01N\\x86\\x94\\x8c\\x06a5_gtf\\x94K\\x02N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh\\x18\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\x1e)}\\x94\\x8c\\x05_name\\x94h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\x10h\\nh\\x12h\\x0bh\\x14h\\x0cub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c.figures/athaliana/athaliana_gurag_seqlogos.svg\\x94\\x8c-figures/athaliana/athaliana_aggu_seqlogos.svg\\x94\\x8c+figures/athaliana/athaliana_u5_u6_ecdfs.svg\\x94\\x8c7figures/athaliana/athaliana_u5_u6_altsplice_scatter.svg\\x94e}\\x94(h\\x0e}\\x94(\\x8c\\ngurag_logo\\x94K\\x00N\\x86\\x94\\x8c\\taggu_logo\\x94K\\x01N\\x86\\x94\\x8c\\x0bu5_u6_ecdfs\\x94K\\x02N\\x86\\x94\\x8c\\x17u5_u6_altsplice_scatter\\x94K\\x03N\\x86\\x94uh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh2h,h4h-h6h.h8h/ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\x0e}\\x94h\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\tathaliana\\x94a}\\x94(h\\x0e}\\x94\\x8c\\x08organism\\x94K\\x00N\\x86\\x94sh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94b\\x8c\\x08organism\\x94hVub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\x0e}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bhmK\\x01hoK\\x01hqhjub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c=notebook_processed/athaliana_u5_u6_splice_site_logos.py.ipynb\\x94a}\\x94(h\\x0e}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x16]\\x94(h\\x18h\\x19eh\\x18h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x18sNt\\x94bh\\x19h\\x1ch\\x1e\\x85\\x94R\\x94(h\\x1e)}\\x94h\"h\\x19sNt\\x94bh\\x83h\\x80ub\\x8c\\x06config\\x94}\\x94\\x8c\\x07genomes\\x94}\\x94(\\x8c\\tathaliana\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8cA../annotations/Araport11_GFF3_genes_transposons.201606.no_chr.gtf\\x94u\\x8c\\x08hsapiens\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c=../annotations/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8c)../annotations/Homo_sapiens.GRCh38.98.gtf\\x94u\\x8c\\x08celegans\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c>../annotations/Caenorhabditis_elegans.WBcel235.dna.toplevel.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8c6../annotations/Caenorhabditis_elegans.WBcel235.104.gtf\\x94u\\x8c\\rdmelanogaster\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c?../annotations/Drosophila_melanogaster.BDGP6.32.dna.toplevel.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8c7../annotations/Drosophila_melanogaster.BDGP6.32.104.gtf\\x94u\\x8c\\x06drerio\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c1../annotations/Danio_rerio.GRCz11.dna.toplevel.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8c)../annotations/Danio_rerio.GRCz11.104.gtf\\x94uus\\x8c\\x04rule\\x94\\x8c\\x1agenerate_splice_site_logos\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cS/cluster/ggs_lab/mtparker/papers/fiona/global_5ss_analysis/rules/notebook_templates\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/cluster/ggs_lab/mtparker/papers/fiona/global_5ss_analysis/pipeline');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import re\n",
    "from glob import glob\n",
    "import random\n",
    "import itertools as it\n",
    "from collections import defaultdict, Counter\n",
    "from functools import reduce, partial\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap, Normalize\n",
    "import matplotlib_logo as logo\n",
    "from matplotlib_logo.logo import calculate_normalised_counts, ALPHABETS\n",
    "import seaborn as sns\n",
    "\n",
    "import pysam\n",
    "\n",
    "\n",
    "## Default plotting params\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set(font='Arial')\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "style = sns.axes_style('white')\n",
    "style.update(sns.axes_style('ticks'))\n",
    "style['xtick.major.size'] = 1\n",
    "style['ytick.major.size'] = 1\n",
    "sns.set(font_scale=1.2, style=style)\n",
    "pal = sns.color_palette(['#0072b2', '#d55e00', '#009e73', '#f0e442', '#cc79a7', '#56b4e9', '#e69f00'])\n",
    "cmap = ListedColormap(pal.as_hex())\n",
    "sns.set_palette(pal)\n",
    "sns.palplot(pal)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RC = str.maketrans('ACGT', 'TGCA')\n",
    "\n",
    "\n",
    "def rev_comp(seq):\n",
    "    return seq.translate(RC)[::-1]\n",
    "\n",
    "\n",
    "def get_gtf_attribute(gtf_record, attribute):\n",
    "    try:\n",
    "        attr = re.search(f'{attribute} \"(.+?)\";', gtf_record[8]).group(1)\n",
    "    except AttributeError:\n",
    "        raise ValueError(\n",
    "            f'Could not parse attribute {attribute} '\n",
    "            f'from GTF with feature type {record[2]}'\n",
    "        )\n",
    "    return attr\n",
    "\n",
    "\n",
    "def gtf_parser(gtf_fn):\n",
    "    gtf_records = {}\n",
    "    with open(gtf_fn) as gtf:\n",
    "        for i, record in enumerate(gtf):\n",
    "            if record.startswith('#') or record.startswith('track'):\n",
    "                continue\n",
    "            record = record.split('\\t')\n",
    "            chrom, _, feat_type, start, end, _, strand = record[:7]\n",
    "            start = int(start) - 1\n",
    "            end = int(end)\n",
    "            if feat_type == 'exon':\n",
    "                transcript_id = get_gtf_attribute(record, 'transcript_id')\n",
    "                gene_id = get_gtf_attribute(record, 'gene_id')\n",
    "                idx = (chrom, gene_id, transcript_id, strand)\n",
    "                if idx not in gtf_records:\n",
    "                    gtf_records[idx] = []\n",
    "                gtf_records[idx].append([start, end])\n",
    "    for (chrom, gene_id, transcript_id, strand), invs in gtf_records.items():\n",
    "        invs.sort()\n",
    "        exons = np.array(invs)\n",
    "        yield chrom, gene_id, transcript_id, strand, exons\n",
    "\n",
    "\n",
    "def get_all_intron_donor_acceptor_seqs(gtf_fn, fasta_fn, winsize=6):\n",
    "    donor_acceptor_seqs = {}\n",
    "    with pysam.FastaFile(fasta_fn) as fasta:\n",
    "        for chrom, gene_id, transcript_id, strand, exons in gtf_parser(gtf_fn):\n",
    "            donors = exons[:-1, 1]\n",
    "            acceptors = exons[1:, 0]\n",
    "            if strand == '-':\n",
    "                donors, acceptors = acceptors, donors\n",
    "            for donor, acceptor in zip(donors, acceptors):\n",
    "                dstr = f'{chrom}:{donor}{strand}'\n",
    "                astr = f'{chrom}:{acceptor}{strand}'\n",
    "                ds = fasta.fetch(chrom, donor - winsize, donor + winsize)\n",
    "                as_ = fasta.fetch(chrom, acceptor - winsize, acceptor + winsize)\n",
    "                if strand == '-':\n",
    "                    ds = rev_comp(ds)\n",
    "                    as_ = rev_comp(as_)\n",
    "                donor_acceptor_seqs[(dstr, astr)] = [\n",
    "                    ds.upper().replace('T', 'U'),\n",
    "                    as_.upper().replace('T', 'U'),\n",
    "                    abs(donor - acceptor)\n",
    "                ]\n",
    "    donor_seqs, acceptor_seqs, intron_lengths = zip(*donor_acceptor_seqs.values())\n",
    "    return donor_seqs, acceptor_seqs, intron_lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RC = str.maketrans('ACGT', 'TGCA')\n",
    "\n",
    "\n",
    "def rev_comp(seq):\n",
    "    return seq.translate(RC)[::-1]\n",
    "\n",
    "\n",
    "IUPAC = {\n",
    "    'A': 'A', 'C': 'C', 'G': 'G', 'U': 'U',\n",
    "    'S': 'GC', 'W': 'AU', 'R': 'AG', 'Y': 'CU',\n",
    "    'K': 'GU', 'M': 'AC', 'N': 'ACGU',\n",
    "}\n",
    "\n",
    "IUPAC_INV = {\n",
    "    'A': 'B', 'C': 'D', 'G': 'H', 'U': 'V',\n",
    "    'S': 'W', 'W': 'S', 'R': 'Y', 'Y': 'R',\n",
    "    'K': 'M', 'M': 'K',\n",
    "    'N': 'N'\n",
    "}\n",
    "\n",
    "\n",
    "def iupac_classify(seq, consensus):\n",
    "    clss = []\n",
    "    for o, e in zip(seq, consensus):\n",
    "        if o in IUPAC[e]:\n",
    "            clss.append(e)\n",
    "        \n",
    "        else:\n",
    "            clss.append(IUPAC_INV[e])\n",
    "    return ''.join(clss)\n",
    "\n",
    "\n",
    "def u5_classify(donor_seq):\n",
    "    assert not len(donor_seq) % 2\n",
    "    ws = len(donor_seq) // 2\n",
    "    return iupac_classify(donor_seq[ws - 2: ws], 'AG')\n",
    "\n",
    "\n",
    "def u6_classify(donor_seq):\n",
    "    assert not len(donor_seq) % 2\n",
    "    ws = len(donor_seq) // 2\n",
    "    return iupac_classify(donor_seq[ws + 2: ws + 5], 'RAG')\n",
    "\n",
    "\n",
    "def acceptor_classify(acceptor_seq):\n",
    "    assert not len(acceptor_seq) % 2\n",
    "    ws = len(acceptor_seq) // 2\n",
    "    return iupac_classify(acceptor_seq[ws - 5: ws - 2], 'UKC')\n",
    "\n",
    "\n",
    "def edit_distance(seq1, seq2):\n",
    "    ed = 0\n",
    "    for i, j in zip(seq1, seq2):\n",
    "        if i != j:\n",
    "            ed += 1\n",
    "    return ed\n",
    "\n",
    "\n",
    "IUPAC_OHE = {\n",
    "    'A': [1, 0, 0, 0],\n",
    "    'C': [0, 1, 0, 0],\n",
    "    'G': [0, 0, 1, 0],\n",
    "    'U': [0, 0, 0, 1],\n",
    "    'R': [0.5, 0, 0.5, 0],\n",
    "    'Y': [0, 0.5, 0, 0.5],\n",
    "    'S': [0, 0.5, 0.5, 0],\n",
    "    'W': [0.5, 0, 0, 0.5],\n",
    "    'B': [0, 1/3, 1/3, 1/3],\n",
    "    'H': [1/3, 1/3, 0, 1/3],\n",
    "    'N': [0.25, 0.25, 0.25, 0.25]\n",
    "}\n",
    "\n",
    "\n",
    "def ohe_seq_iupac(seq):\n",
    "    seq_ohe = []\n",
    "    for n in seq:\n",
    "        seq_ohe.append(IUPAC_OHE[n])\n",
    "    return np.array(seq_ohe)\n",
    "\n",
    "\n",
    "def pwm_score(seq, pwm):\n",
    "    pwm = np.array(pwm)\n",
    "    seq_ohe = ohe_seq_iupac(seq)\n",
    "    return (seq_ohe * pwm).ravel().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_donor_acceptor_logos(seqs, seq_type='donor', title=None, ax=None):\n",
    "    w = len(seqs[0]) // 2\n",
    "    if seq_type == 'donor':\n",
    "        trim_method = lambda seqs: [s[w - 2: w + 5] for s in seqs]\n",
    "        xticks = np.arange(7) + 0.5\n",
    "        xticklabels = ['−2', '−1', '+1', '+2', '+3', '+4', '+5']\n",
    "    elif seq_type == 'acceptor':\n",
    "        trim_method = lambda seqs: [s[w - 5: w + 2] for s in seqs]\n",
    "        xticks = np.arange(7) + 0.5\n",
    "        xticklabels = ['−5', '−4', '−3', '−2', '−1', '+1', '+2']\n",
    "    else:\n",
    "        trim_method = lambda seqs: seqs\n",
    "        xticks = []\n",
    "        xticklabels = []\n",
    "    ax = logo.draw_logo(\n",
    "        trim_method(seqs),\n",
    "        alphabet='rna',\n",
    "        y_format='probability',\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.set_xticks(xticks)\n",
    "    ax.tick_params(bottom=False)\n",
    "    ax.set_xticklabels(xticklabels)\n",
    "    if title is not None:\n",
    "        ax.set_title(f'{title} (n={len(seqs)})')\n",
    "    return ax\n",
    "\n",
    "\n",
    "def all_u5_u6_classifications(fasta_fn, gtf_fn):\n",
    "    print('loading donor/acceptor seqs')\n",
    "    donor_seqs, acceptor_seqs, intron_lengths = get_all_intron_donor_acceptor_seqs(\n",
    "        gtf_fn,\n",
    "        fasta_fn\n",
    "    )\n",
    "    ax = plot_donor_acceptor_logos(donor_seqs)\n",
    "    plt.show()\n",
    "    ax = plot_donor_acceptor_logos(acceptor_seqs, seq_type='acceptor')\n",
    "    plt.show()\n",
    "    donor_pwm = np.log(calculate_normalised_counts(donor_seqs, ALPHABETS['rna']))\n",
    "    pwms = {'U5': donor_pwm[4:6], 'U6': donor_pwm[8:11]}\n",
    "\n",
    "    u5_u6_classifications = []\n",
    "\n",
    "    for seq, ilen in zip(donor_seqs, intron_lengths):\n",
    "        u5_u6_classifications.append(\n",
    "            [seq, \n",
    "             u5_classify(seq), pwm_score(seq[4:6], pwms['U5']),\n",
    "             u6_classify(seq), pwm_score(seq[8:11], pwms['U6']),\n",
    "             ilen]\n",
    "        )\n",
    "\n",
    "    u5_u6_classifications = pd.DataFrame(\n",
    "        u5_u6_classifications,\n",
    "        columns=['seq', 'u5_class', 'u5_pwm_score',\n",
    "                 'u6_class', 'u6_pwm_score', 'intron_length']\n",
    "    )\n",
    "    print('U5 vs U6 score correlation')\n",
    "    print(stats.spearmanr(\n",
    "        u5_u6_classifications.u5_pwm_score,\n",
    "        u5_u6_classifications.u6_pwm_score,\n",
    "    ))\n",
    "    return donor_seqs, acceptor_seqs, u5_u6_classifications, pwms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donors, acceptors, u5_u6_classifications, pwms = all_u5_u6_classifications(\n",
    "    snakemake.input.fasta,\n",
    "    snakemake.input.gtf,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gurag = u5_u6_classifications.query('u6_class == \"RAG\"').seq.tolist()\n",
    "nongurag = u5_u6_classifications.query('u6_class != \"RAG\"').seq.tolist()\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(6.5, 4.5), nrows=2)\n",
    "\n",
    "plot_donor_acceptor_logos(gurag, title='//GURAG 5\\'SSs', ax=axes[0])\n",
    "plot_donor_acceptor_logos(nongurag, title='Non- //GURAG 5\\'SSs', ax=axes[1])\n",
    "plt.tight_layout()\n",
    "plt.savefig(snakemake.output.gurag_logo)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aggu = u5_u6_classifications.query('u5_class == \"AG\"').seq.tolist()\n",
    "nonaggu = u5_u6_classifications.query('u5_class != \"AG\"').seq.tolist()\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(6.5, 4.5), nrows=2)\n",
    "\n",
    "plot_donor_acceptor_logos(aggu, title='AG//GU 5\\'SSs', ax=axes[0])\n",
    "plot_donor_acceptor_logos(nonaggu, title='Non- AG//GU 5\\'SSs', ax=axes[1])\n",
    "plt.tight_layout()\n",
    "plt.savefig(snakemake.output.aggu_logo)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ecdf(data, ax=None, color=None, **plot_kwargs):\n",
    "    x = np.unique(data)\n",
    "    ecdf = np.cumsum([(data == n).sum() for n in x]) / len(data)\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots()\n",
    "    ax.step(x, ecdf, where='post', color=color, **plot_kwargs)\n",
    "    return ax\n",
    "\n",
    "\n",
    "def plot_u6_u5_ecdfs(classifications):\n",
    "    fig, axes = plt.subplots(figsize=(8.5, 4), ncols=2, sharey=True)\n",
    "    plot_ecdf(\n",
    "        classifications.query('u6_class == \"RAG\"').u5_pwm_score,\n",
    "        ax=axes[0],\n",
    "        lw=3,\n",
    "        alpha=0.8,\n",
    "        label='//GURAG',\n",
    "        color=pal[0],\n",
    "    )\n",
    "    plot_ecdf(\n",
    "        classifications.query('u6_class != \"RAG\"').u5_pwm_score,\n",
    "        ax=axes[0],\n",
    "        lw=3,\n",
    "        alpha=0.8,\n",
    "        label='Other',\n",
    "        color=pal[1],\n",
    "    )\n",
    "    axes[0].legend(title='U6 class', loc=2)\n",
    "    axes[0].set_xlabel('U5 PSSM log-likelihood')\n",
    "\n",
    "    for i, u5_cls in enumerate(['AG', 'BG', 'AH', 'BH']):\n",
    "        plot_ecdf(\n",
    "            classifications.query('u5_class == @u5_cls').u6_pwm_score,\n",
    "            ax=axes[1],\n",
    "            lw=3,\n",
    "            alpha=0.8,\n",
    "            label=u5_cls + '//GU',\n",
    "            color=pal[i],\n",
    "        )\n",
    "    axes[1].legend(title='U5 class', loc=2)\n",
    "    axes[1].set_xlabel('U6 PSSM log-likelihood')\n",
    "    return fig, axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_u6_u5_ecdfs(u5_u6_classifications)\n",
    "plt.savefig(snakemake.output.u5_u6_ecdfs)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_annot_alt_seqs(suppa_alt5_gtf_fn, fasta_fn, winsize=6, which='a5'):\n",
    "    seq_pairs = {}\n",
    "    other_seq = {}\n",
    "    with pysam.FastaFile(fasta_fn) as fasta:\n",
    "        for chrom, _, event_id, strand, invs in gtf_parser(suppa_alt5_gtf_fn):\n",
    "            event_id, alt_id = re.split('[.:]alt(?:ernative)?', event_id)\n",
    "            pos = invs[0, 1] if which == 'a5' else invs[1, 0]\n",
    "            opos = invs[1, 0] if which == 'a5' else invs[0, 1]\n",
    "            seq = fasta.fetch(chrom, pos - winsize, pos + winsize).upper()\n",
    "            oseq = fasta.fetch(chrom, opos - winsize, opos + winsize).upper()\n",
    "            if strand == '-':\n",
    "                seq = rev_comp(oseq)\n",
    "                oseq = rev_comp(seq)\n",
    "            if event_id not in seq_pairs:\n",
    "                seq_pairs[event_id] = []\n",
    "            seq_pairs[event_id].append([seq.replace('T', 'U'), pos, strand])\n",
    "            other_seq[event_id] = oseq.replace('T', 'U')\n",
    "    # order by more upstream\n",
    "    seq_pairs_ordered = []\n",
    "    other_seq_ordered = []\n",
    "    for event_id, pairs in seq_pairs.items():\n",
    "        seq1, alt1_pos, strand = pairs.pop()\n",
    "        seq2, alt2_pos, _ = pairs.pop()\n",
    "        if alt1_pos < alt2_pos:\n",
    "            if strand == '+':\n",
    "                seq_pairs_ordered.append([seq1, seq2])\n",
    "            else:\n",
    "                seq_pairs_ordered.append([seq2, seq1])\n",
    "        else:\n",
    "            if strand == '+':\n",
    "                seq_pairs_ordered.append([seq2, seq1])\n",
    "            else:\n",
    "                seq_pairs_ordered.append([seq1, seq2])\n",
    "        other_seq_ordered.append(other_seq[event_id])\n",
    "    return seq_pairs_ordered, other_seq_ordered\n",
    "\n",
    "\n",
    "def alt_donor_pair_scores(a5_gtf, fasta_fn, pwms):\n",
    "    alt_donor_pairs, alt_donor_pair_acceptor = get_annot_alt_seqs(\n",
    "        a5_gtf, fasta_fn\n",
    "    )\n",
    "    u5_u6_classifications = []\n",
    "\n",
    "    for seq1, seq2 in alt_donor_pairs:\n",
    "        u5_u6_classifications.append(\n",
    "            [seq1, seq2,\n",
    "             u5_classify(seq1), pwm_score(seq1[4:6], pwms['U5']),\n",
    "             u5_classify(seq2), pwm_score(seq2[4:6], pwms['U5']),\n",
    "             u6_classify(seq1), pwm_score(seq1[8:11], pwms['U6']),\n",
    "             u6_classify(seq2), pwm_score(seq2[8:11], pwms['U6'])]\n",
    "        )\n",
    "\n",
    "    u5_u6_classifications = pd.DataFrame(\n",
    "        u5_u6_classifications,\n",
    "        columns=['upstream_seq', 'downstream_seq',\n",
    "                 'upstream_u5_class', 'upstream_u5_pwm_score',\n",
    "                 'downstream_u5_class', 'downstream_u5_pwm_score',\n",
    "                 'upstream_u6_class', 'upstream_u6_pwm_score',\n",
    "                 'downstream_u6_class', 'downstream_u6_pwm_score']\n",
    "    )\n",
    "    u5_u6_classifications['u5_log_odds'] = (\n",
    "        u5_u6_classifications.upstream_u5_pwm_score - u5_u6_classifications.downstream_u5_pwm_score\n",
    "    )\n",
    "    u5_u6_classifications['u6_log_odds'] = (\n",
    "        u5_u6_classifications.upstream_u6_pwm_score - u5_u6_classifications.downstream_u6_pwm_score\n",
    "    )\n",
    "    return u5_u6_classifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alt_donor_pair_scatterplot(a5_gtf, fasta_fn, pwms):\n",
    "    \n",
    "    u5_u6_classifications = alt_donor_pair_scores(a5_gtf, fasta_fn, pwms)\n",
    "    fig = plt.figure(figsize=(12.5, 5))\n",
    "    scatter_ax = plt.subplot2grid(shape=(12, 13), loc=(0, 0), rowspan=12, colspan=5)\n",
    "    logo_ax1 = plt.subplot2grid(shape=(12, 13), loc=(0, 6), rowspan=5, colspan=7)\n",
    "    logo_ax2 = plt.subplot2grid(shape=(12, 13), loc=(7, 6), rowspan=5, colspan=7)\n",
    "    \n",
    "    sns.regplot(\n",
    "        x='u5_log_odds',\n",
    "        y='u6_log_odds',\n",
    "        data=u5_u6_classifications,\n",
    "        ax=scatter_ax,\n",
    "        scatter_kws=dict(color=pal[-2], alpha=0.25, s=4, marker='x', rasterized=True),\n",
    "        line_kws=dict(color=pal[0])\n",
    "    )\n",
    "    rho = stats.spearmanr(u5_u6_classifications.u5_log_odds, u5_u6_classifications.u6_log_odds)[0]\n",
    "    scatter_ax.annotate(xy=(0.05, 0.05), s=f'Spearmans rho={rho:.2f}', xycoords='axes fraction')\n",
    "    scatter_ax.set_xlabel('U5 PSSM log odds')\n",
    "    scatter_ax.set_ylabel('U6 PSSM log odds')\n",
    "\n",
    "    u6_stronger = []\n",
    "    u6_weaker = []\n",
    "    for _, ups, dwns, u6d in u5_u6_classifications[['upstream_seq', 'downstream_seq', 'u6_log_odds']].itertuples():\n",
    "        if u6d < 0:\n",
    "            u6_stronger.append(dwns)\n",
    "            u6_weaker.append(ups)\n",
    "        elif u6d >= 0:\n",
    "            u6_stronger.append(ups)\n",
    "            u6_weaker.append(dwns)\n",
    "\n",
    "    plot_donor_acceptor_logos(u6_stronger, title='Alt 5\\'SS with higher U6 PSSM score', ax=logo_ax1)\n",
    "    plot_donor_acceptor_logos(u6_weaker, title='Alt 5\\'SS with lower U6 PSSM score', ax=logo_ax2)\n",
    "\n",
    "    return fig, [scatter_ax, logo_ax1, logo_ax2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = alt_donor_pair_scatterplot(\n",
    "    snakemake.input.a5_gtf,\n",
    "    snakemake.input.fasta,\n",
    "    pwms\n",
    ")\n",
    "plt.savefig(snakemake.output.u5_u6_altsplice_scatter)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
