{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/cluster/ggs_lab/mtparker/.conda/envs/snakemake6/lib/python3.10/site-packages', '/cluster/ggs_lab/mtparker/papers/fiona/fiona_temp_rnaseq/rules/notebook_templates']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\xba\\x08\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c\\x1bsplicing/denovo/psi_fit.csv\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x07psi_fit\\x94K\\x00N\\x86\\x94\\x8c\\x0fgenome_fasta_fn\\x94K\\x01N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh\\x15\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\x1b)}\\x94\\x8c\\x05_name\\x94h\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bh\\x0fh\\nh\\x11h\\x0bub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8c-figures/splicing/denovo/A5_logos_and_heatmaps\\x94a}\\x94(h\\r}\\x94\\x8c\\x0csplice_plots\\x94K\\x00N\\x86\\x94sh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bh,h)ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\r}\\x94h\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94(\\x8c\\x06denovo\\x94\\x8c\\x02A5\\x94e}\\x94(h\\r}\\x94(\\x8c\\x08ref_type\\x94K\\x00N\\x86\\x94\\x8c\\nevent_type\\x94K\\x01N\\x86\\x94uh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94b\\x8c\\x08ref_type\\x94hJ\\x8c\\nevent_type\\x94hKub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\r}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bheK\\x01hgK\\x01hihbub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c4notebook_processed/sequence_logos_denovo_A5.py.ipynb\\x94a}\\x94(h\\r}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bh{hxub\\x8c\\x06config\\x94}\\x94(\\x8c\\x0fgenome_fasta_fn\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94\\x8c\\x0cannot_gtf_fn\\x94\\x8cA../annotations/Araport11_GFF3_genes_transposons.201606.no_chr.gtf\\x94\\x8c\\x16transcriptome_fasta_fn\\x94]\\x94(\\x8c0../annotations/Araport11_genes.201606.cdna.fasta\\x94\\x8c5../annotations/Araport11_non_coding.201606.cdna.fasta\\x94e\\x8c\\x11proteome_fasta_fn\\x94\\x8c4../annotations/Araport11_genes.201606.pep.repr.fasta\\x94\\x8c\\x14fastq_screen_genomes\\x94}\\x94(\\x8c\\tathaliana\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94\\x8c\\x08hsapiens\\x94\\x8c=../annotations/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa\\x94\\x8c\\x07at_rrna\\x94\\x8c\\x1c../annotations/arath_rrna.fa\\x94\\x8c\\x05ecoli\\x94\\x8c\"../annotations/ecoli_k12_mg1655.fa\\x94u\\x8c\\x11suppa_comparisons\\x94]\\x94(\\x8c\\x14col0_04c_vs_col0_20c\\x94\\x8c\\x14col0_12c_vs_col0_20c\\x94\\x8c\\x14col0_28c_vs_col0_20c\\x94\\x8c\\x14fio1_04c_vs_fio1_20c\\x94\\x8c\\x14fio1_12c_vs_fio1_20c\\x94\\x8c\\x14fio1_28c_vs_fio1_20c\\x94\\x8c\\x14fio1_04c_vs_col0_04c\\x94\\x8c\\x14fio1_12c_vs_col0_12c\\x94\\x8c\\x14fio1_20c_vs_col0_20c\\x94\\x8c\\x14fio1_28c_vs_col0_28c\\x94eu\\x8c\\x04rule\\x94\\x8c\\x1bgenerate_logos_and_heatmaps\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cQ/cluster/ggs_lab/mtparker/papers/fiona/fiona_temp_rnaseq/rules/notebook_templates\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/cluster/ggs_lab/mtparker/papers/fiona/fiona_temp_rnaseq/pipeline');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "import itertools as it\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap, Normalize\n",
    "import seaborn as sns\n",
    "from IPython.display import Markdown, display_markdown\n",
    "\n",
    "from seqlogo_utils import (\n",
    "    iupac_classify,\n",
    "    u5_classify,\n",
    "    get_donor_acceptor_seqs_from_df,\n",
    "    plot_donor_acceptor_logos,\n",
    "    plot_u5_u6_heatmap,\n",
    "    calculate_switch_nt,\n",
    "    perc_seq_pos,\n",
    "    perc_seq_switch_pos,\n",
    "    get_pairwise_seqs,\n",
    "    g_test_seqs,\n",
    ")\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set(font='Arial')\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "style = sns.axes_style('white')\n",
    "style.update(sns.axes_style('ticks'))\n",
    "style['xtick.major.size'] = 1\n",
    "style['ytick.major.size'] = 1\n",
    "sns.set(font_scale=1.2, style=style)\n",
    "pal = sns.color_palette(['#0072b2', '#d55e00', '#009e73', '#f0e442', '#cc79a7', '#56b4e9', '#e69f00'])\n",
    "cmap = ListedColormap(pal.as_hex())\n",
    "sns.set_palette(pal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_PATH = snakemake.output.splice_plots\n",
    "if not os.path.exists(OUTPUT_PATH):\n",
    "    os.makedirs(OUTPUT_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_formatted_markdown(md, **kwargs):\n",
    "    md_f = md.format(**kwargs)\n",
    "    display_markdown(Markdown(md_f))\n",
    "\n",
    "MD_TEXT = open('../rules/notebook_templates/md_text/A5.md').readlines()\n",
    "\n",
    "ref_type = {'ref': 'reference', 'denovo': '*de novo*'}\n",
    "\n",
    "display_formatted_markdown(\n",
    "    '# Alternative 5\\'SS analysis using {ref_type} transcriptome assembly',\n",
    "    ref_type=ref_type[snakemake.wildcards.ref_type]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_coords(coords):\n",
    "    if not coords:\n",
    "        return None\n",
    "    else:\n",
    "        with warnings.catch_warnings():\n",
    "            warnings.simplefilter(\"ignore\")\n",
    "            return np.fromstring(coords.strip('[] '), sep=' ', dtype=int)\n",
    "\n",
    "psi_fit = pd.read_csv(\n",
    "    snakemake.input.psi_fit,\n",
    "    converters={'chrom': str,\n",
    "                'alt1': convert_coords,\n",
    "                'alt2': convert_coords}\n",
    ")\n",
    "psi_fit_sig = psi_fit.query(\n",
    "    'event_class == \"A5\" & (geno_fdr < 0.05 | gxt_fdr < 0.05)'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_acceptor_pos, donor_acceptor_seqs = get_donor_acceptor_seqs_from_df(\n",
    "    psi_fit_sig, snakemake.input.genome_fasta_fn, winsize=6\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a5_pos, a5_seqs = get_donor_acceptor_seqs_from_df(\n",
    "    psi_fit_sig, snakemake.input.genome_fasta_fn, winsize=6\n",
    ")\n",
    "\n",
    "a5_alt1_donors, a5_alt1_acceptors, a5_alt2_donors, a5_alt2_acceptors = (\n",
    "    list(a5_seqs['alt1']['donor'].values()),\n",
    "    list(a5_seqs['alt1']['acceptor'].values()),\n",
    "    list(a5_seqs['alt2']['donor'].values()),\n",
    "    list(a5_seqs['alt2']['acceptor'].values()),\n",
    ")\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[0],\n",
    "    num_a5=len(psi_fit_sig),\n",
    "    g_test_p=g_test_seqs(*get_pairwise_seqs(a5_pos, a5_seqs, 'donor'))[1],\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(15, 7))\n",
    "alt1_logo_ax = plt.subplot2grid((3, 2), (0, 0))\n",
    "alt1_hmap_ax = plt.subplot2grid((3, 2), (1, 0), rowspan=2)\n",
    "alt2_logo_ax = plt.subplot2grid((3, 2), (0, 1))\n",
    "alt2_hmap_ax = plt.subplot2grid((3, 2), (1, 1), rowspan=2)\n",
    "\n",
    "plot_donor_acceptor_logos(\n",
    "    a5_alt1_donors,\n",
    "    seq_type='donor',\n",
    "    title='fio1-3 sensitive 5\\'SS',\n",
    "    ax=alt1_logo_ax\n",
    ")\n",
    "_, a5_alt1_u5_grp, a5_alt1_u6_grp = plot_u5_u6_heatmap(a5_alt1_donors, vmin=0, vmax=616, ax=alt1_hmap_ax)\n",
    "\n",
    "plot_donor_acceptor_logos(\n",
    "    a5_alt2_donors,\n",
    "    seq_type='donor',\n",
    "    title='Alt. 5\\'SS increased in fio1-3',\n",
    "    ax=alt2_logo_ax\n",
    ")\n",
    "_, a5_alt2_u5_grp, a5_alt2_u6_grp = plot_u5_u6_heatmap(a5_alt2_donors, vmin=0, vmax=616, ax=alt2_hmap_ax)\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_fio1_logo.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# +4 position A vs not A contigency table\n",
    "a4 = np.array([g[1] for g in a5_alt1_u6_grp + a5_alt2_u6_grp])\n",
    "y = np.repeat([1, 0], [len(a5_alt1_u6_grp), len(a5_alt2_u6_grp)])\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[1],\n",
    "    a4_acc=100 - np.mean((a4 == 'A')  ^ (y == 1)) * 100\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 3))\n",
    "sns.heatmap(\n",
    "    pd.crosstab(y, a4),\n",
    "    ax=ax,\n",
    "    cmap='Blues',\n",
    "    square=True,\n",
    "    cbar=False,\n",
    "    annot=True,\n",
    "    fmt='d'\n",
    ")\n",
    "ax.set_xlabel('Base at +4 position')\n",
    "ax.set_yticklabels(['Increase', 'Decrease'], va='center')\n",
    "ax.set_ylabel('ΔPSI in fio1-3')\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_u6_contigency_table_p4.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# full contingency table for +4 position\n",
    "a5_alt1_pairwise, a5_alt2_pairwise = get_pairwise_seqs(a5_pos, a5_seqs)\n",
    "a4_alt1 = np.array([g[9] for g in a5_alt1_pairwise])\n",
    "a4_alt2 = np.array([g[9] for g in a5_alt2_pairwise])\n",
    "\n",
    "ab_mask = perc_seq_switch_pos(a5_alt1_pairwise, a5_alt2_pairwise, 4, 'A', 'B')[0]\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[2],\n",
    "    a5_perc_u4=perc_seq_pos(a5_alt2_donors, 4, 'U')[1],\n",
    "    a5_perc_ab4=perc_seq_switch_pos(a5_alt1_pairwise, a5_alt2_pairwise, 4, 'A', 'B')[1],\n",
    "    a5_perc_au4=perc_seq_pos(\n",
    "        [s for s, m in zip(a5_alt2_pairwise, ab_mask) if m],\n",
    "        4, 'U'\n",
    "    )[1],\n",
    "    a5_perc_ba4=perc_seq_switch_pos(a5_alt1_pairwise, a5_alt2_pairwise, 4, 'B', 'A')[1],\n",
    "    a5_perc_su4=perc_seq_switch_pos(a5_alt1_pairwise, a5_alt2_pairwise, 4, 'S', 'U')[1],\n",
    "    a5_perc_nn4=np.mean(a4_alt1 == a4_alt2) * 100\n",
    ")\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(10, 5), ncols=2)\n",
    "sns.heatmap(\n",
    "    pd.crosstab(a4_alt2, a4_alt1),\n",
    "    ax=axes[0],\n",
    "    cmap='Blues',\n",
    "    square=True,\n",
    "    cbar=False,\n",
    "    annot=True,\n",
    "    fmt='d'\n",
    ")\n",
    "axes[0].set_xlabel('Base at +4 position in reduced site')\n",
    "axes[0].set_ylabel('Base at +4 position in increased site')\n",
    "\n",
    "# -1/+5 contigency table for 5'SS pairs with equivalent +4 positions\n",
    "\n",
    "mask = [a1 == a2 for a1, a2 in zip(a4_alt1, a4_alt2)]\n",
    "is_g = partial(iupac_classify, consensus='G')\n",
    "masked_alt1_m1p5 = [f'{is_g(seq[5])}/{is_g(seq[10])}' for seq, m\n",
    "                    in zip(a5_alt1_pairwise, mask) if m]\n",
    "masked_alt2_m1p5 = [f'{is_g(seq[5])}/{is_g(seq[10])}' for seq, m\n",
    "                    in zip(a5_alt2_pairwise, mask) if m]\n",
    "\n",
    "\n",
    "contingency_table = pd.crosstab(\n",
    "        pd.Series(masked_alt2_m1p5, name='Increased 5\\'SS -1/+5 base'),\n",
    "        pd.Series(masked_alt1_m1p5, name='Reduced 5\\'SS -1/+5 base'),\n",
    ")\n",
    "contingency_table = contingency_table.loc[\n",
    "    ('H/G', 'H/H', 'G/G', 'G/H'),\n",
    "    ('H/G', 'G/G', 'H/H', 'G/H'),\n",
    "]\n",
    "\n",
    "sns.heatmap(\n",
    "    contingency_table,\n",
    "    ax=axes[1],\n",
    "    cmap='Blues',\n",
    "    square=True,\n",
    "    cbar=False,\n",
    "    annot=True,\n",
    "    fmt='d'\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_u6_contigency_table_full.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_paired_base_effect_size(alt1_seqs, alt2_seqs, pos, dpsi,\n",
    "                                 order, pal, iupac_filt,\n",
    "                                 ax=None):\n",
    "    # convert 5'SS pos to index\n",
    "    i = len(alt1_seqs[0]) // 2 + pos - 1\n",
    "    alt1_base = [iupac_filt[seq[i]] for seq in alt1_seqs]\n",
    "    alt2_base = [iupac_filt[seq[i]] for seq in alt2_seqs]\n",
    "    base_grp = [f'{a1}/{a2}' for a1, a2 in zip(alt1_base, alt2_base)]\n",
    "\n",
    "    mirrored_order = ['/'.join(o) for o in (it.product(order, order[::-1]))]\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(len(mirrored_order) / 1.5, 4))\n",
    "    sns.boxplot(\n",
    "        x=base_grp,\n",
    "        y=dpsi,\n",
    "        showfliers=False,\n",
    "        order=mirrored_order,\n",
    "        palette=[pal[o[0]] for o in mirrored_order],\n",
    "        ax=ax\n",
    "    )\n",
    "    ax.set_xlabel(f'{pos:+} base of fio1-3 sensitive/\\nalternative 5\\'SS')\n",
    "    ax.set_ylabel('Effect size (absolute ΔPSI)')\n",
    "    return ax\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[3]\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(11, 5))\n",
    "p4_ax = plt.subplot2grid((1, 10), (0, 0), colspan=6)\n",
    "p5_ax = plt.subplot2grid((1, 10), (0, 6), colspan=4, sharey=p4_ax)\n",
    "\n",
    "plot_paired_base_effect_size(\n",
    "    a5_alt1_pairwise, a5_alt2_pairwise, 4,\n",
    "    psi_fit_sig.dpsi,\n",
    "    order='ASU',\n",
    "    pal={'A': pal[2], 'S': '#888888', 'U': pal[1]},\n",
    "    iupac_filt={'A': 'A', 'U': 'U', 'G': 'S', 'C': 'S'},\n",
    "    ax=p4_ax\n",
    ")\n",
    "\n",
    "plot_paired_base_effect_size(\n",
    "    a5_alt1_pairwise, a5_alt2_pairwise, 5,\n",
    "    psi_fit_sig.dpsi,\n",
    "    order='GH',\n",
    "    pal={'H': '#888888', 'G': pal[3]},\n",
    "    iupac_filt={'A': 'H', 'U': 'H', 'G': 'G', 'C': 'H'},\n",
    "    ax=p5_ax\n",
    ")\n",
    "p4_ax.set_ylim(0, 0.5)\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_u6_base_effect_size.svg'))\n",
    "plt.show()\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[4]\n",
    ")\n",
    "\n",
    "order = ['AG', 'BG', 'AH', 'BH']\n",
    "fig, ax = plt.subplots(figsize=(4, 4))\n",
    "sns.boxplot(\n",
    "    x=[u5_classify(s) for s in a5_alt1_pairwise],\n",
    "    y=psi_fit_sig.dpsi,\n",
    "    showfliers=False,\n",
    "    order=order,\n",
    "    color=pal[0]\n",
    ")\n",
    "ax.set_ylim(0, 0.5)\n",
    "ax.set_xlabel('U5 class of fio1-3 sensitive 5\\'SS')\n",
    "ax.set_ylabel('Effect size (absolute ΔPSI)')\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_u5_class_effect_size.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a5_switch = psi_fit_sig.apply(calculate_switch_nt, axis=1)\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[5]\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 5))\n",
    "ax.hist(a5_switch, bins=np.arange(-80, 81), align='left')\n",
    "ax.set_ylabel('Alt5\\'SS pairs')\n",
    "ax.set_xlabel('Distance between alternative 5\\'SS pairs with reduced/increased usage in fio1-3')\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_switch_size_histogram.svg'))\n",
    "plt.show()\n",
    "\n",
    "_, a5_temp_seqs = get_donor_acceptor_seqs_from_df(\n",
    "    psi_fit_sig[a5_switch == 5],\n",
    "    snakemake.input.genome_fasta_fn, winsize=7\n",
    ")\n",
    "plot_donor_acceptor_logos(\n",
    "    [s[:12] for s in a5_temp_seqs['alt2']['donor'].values()],\n",
    "    seq_type=None,\n",
    "    title='Alt 5\\'SS pairs with switch distance of +5'\n",
    ")\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_5nt_switch_logo.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Temperature alt 5'SS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a5_temp_pos_1, a5_temp_seqs_1 = get_donor_acceptor_seqs_from_df(\n",
    "    psi_fit.query('event_class == \"A5\" & (temp_fdr < 0.05 | txt_fdr < 0.05) & dpsi2 < 0'),\n",
    "    snakemake.input.genome_fasta_fn, winsize=6\n",
    ")\n",
    "a5_temp_pos_2, a5_temp_seqs_2 = get_donor_acceptor_seqs_from_df(\n",
    "    psi_fit.query('event_class == \"A5\" & (temp_fdr < 0.05 | txt_fdr < 0.05) & dpsi2 > 0'),\n",
    "    snakemake.input.genome_fasta_fn, winsize=6\n",
    ")\n",
    "# merge pos and neg dpsis\n",
    "a5_temp_pos = {\n",
    "    'alt1': {'donor': a5_temp_pos_1['alt1']['donor'] + a5_temp_pos_2['alt2']['donor']},\n",
    "    'alt2': {'donor': a5_temp_pos_1['alt2']['donor'] + a5_temp_pos_2['alt1']['donor']},\n",
    "}\n",
    "a5_temp_seqs = {\n",
    "    'alt1': {'donor': {**a5_temp_seqs_1['alt1']['donor'], **a5_temp_seqs_2['alt2']['donor']}},\n",
    "    'alt2': {'donor': {**a5_temp_seqs_1['alt2']['donor'], **a5_temp_seqs_2['alt1']['donor']}},\n",
    "}\n",
    "cold_a5_donors = list(a5_temp_seqs['alt1']['donor'].values())\n",
    "hot_a5_donors = list(a5_temp_seqs['alt2']['donor'].values())\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[6],\n",
    "    g_test_p=g_test_seqs(*get_pairwise_seqs(a5_temp_pos, a5_temp_seqs))[1]\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(15, 7))\n",
    "cold_logo_ax = plt.subplot2grid((3, 2), (0, 0))\n",
    "cold_hmap_ax = plt.subplot2grid((3, 2), (1, 0), rowspan=2)\n",
    "hot_logo_ax = plt.subplot2grid((3, 2), (0, 1))\n",
    "hot_hmap_ax = plt.subplot2grid((3, 2), (1, 1), rowspan=2)\n",
    "\n",
    "\n",
    "plot_donor_acceptor_logos(\n",
    "    cold_a5_donors,\n",
    "    seq_type='donor',\n",
    "    title='Alt. 5\\'SS used more at low temperature',\n",
    "    ax=cold_logo_ax\n",
    ")\n",
    "plot_u5_u6_heatmap(cold_a5_donors, ax=cold_hmap_ax, vmin=0, vmax=139)\n",
    "plot_donor_acceptor_logos(\n",
    "    hot_a5_donors,\n",
    "    seq_type='donor',\n",
    "    title='Alt. 5\\'SS used more at high temperature',\n",
    "    ax=hot_logo_ax\n",
    ")\n",
    "plot_u5_u6_heatmap(hot_a5_donors, ax=hot_hmap_ax, vmin=0, vmax=139)\n",
    "plt.savefig(os.path.join(OUTPUT_PATH, 'a5_temp_logo_heatmap.svg'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
