{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/cluster/ggs_lab/mtparker/.conda/envs/snakemake6/lib/python3.10/site-packages', '/cluster/ggs_lab/mtparker/papers/fiona/fiona_nanopore/rules/notebook_templates']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\x84\\t\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c\"yanocomp/fip37_vs_fio1_vs_col0.bed\\x94\\x8c1yanocomp/motif_detection/fip37_vs_col0.motifs.bed\\x94\\x8c\"../annotations/miclip_peaks.bed.gz\\x94\\x8cA../annotations/Araport11_GFF3_genes_transposons.201606.no_chr.gtf\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x0eyanocomp_sites\\x94K\\x00K\\x01\\x86\\x94\\x8c\\x0fyanocomp_motifs\\x94K\\x01K\\x02\\x86\\x94\\x8c\\x0cmiclip_peaks\\x94K\\x02N\\x86\\x94\\x8c\\x03gtf\\x94K\\x03N\\x86\\x94\\x8c\\x05fasta\\x94K\\x04N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh\\x1e\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h$)}\\x94\\x8c\\x05_name\\x94h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bh\\x12h\\x06\\x8c\\tNamedlist\\x94\\x93\\x94)\\x81\\x94h\\na}\\x94(h\\x10}\\x94h\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bubh\\x14h/)\\x81\\x94h\\x0ba}\\x94(h\\x10}\\x94h\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bubh\\x16h\\x0ch\\x18h\\rh\\x1ah\\x0eub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c\\'figures/yanocomp/yanocomp_site_logo.svg\\x94\\x8c(figures/yanocomp/yanocomp_site_upset.svg\\x94\\x8c7figures/yanocomp/yanocomp_site_feature_distribution.svg\\x94\\x8c.figures/yanocomp/yanocomp_site_effect_size.svg\\x94e}\\x94(h\\x10}\\x94(\\x8c\\tsite_logo\\x94K\\x00N\\x86\\x94\\x8c\\nsite_upset\\x94K\\x01N\\x86\\x94\\x8c\\x0csite_distrib\\x94K\\x02N\\x86\\x94\\x8c\\x10site_effect_size\\x94K\\x03N\\x86\\x94uh\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bhRhLhThMhVhNhXhOub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\x10}\\x94h\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\x10}\\x94h\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\x10}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bh\\x89K\\x01h\\x8bK\\x01h\\x8dh\\x86ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c-notebook_processed/yanocomp_analysis.py.ipynb\\x94a}\\x94(h\\x10}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x1c]\\x94(h\\x1eh\\x1feh\\x1eh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1esNt\\x94bh\\x1fh\"h$\\x85\\x94R\\x94(h$)}\\x94h(h\\x1fsNt\\x94bh\\x9fh\\x9cub\\x8c\\x06config\\x94}\\x94(\\x8c\\x16transcriptome_fasta_fn\\x94\\x8c0../annotations/Araport11_genes.201606.cdna.fasta\\x94\\x8c\\x0fgenome_fasta_fn\\x94\\x8c:../annotations/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa\\x94\\x8c\\x06gtf_fn\\x94\\x8cA../annotations/Araport11_GFF3_genes_transposons.201606.no_chr.gtf\\x94\\x8c\\x0cmiclip_peaks\\x94\\x8c\"../annotations/miclip_peaks.bed.gz\\x94\\x8c\\x08flowcell\\x94\\x8c\\nFLO-MIN106\\x94\\x8c\\x03kit\\x94\\x8c\\nSQK-RNA002\\x94\\x8c\\x13minimap2_parameters\\x94}\\x94\\x8c\\x0fmax_intron_size\\x94M Ns\\x8c\\x12d3pendr_parameters\\x94}\\x94(\\x8c\\x10min_read_overlap\\x94G?\\xc9\\x99\\x99\\x99\\x99\\x99\\x9a\\x8c\\x06nboots\\x94M\\xe7\\x03\\x8c\\x0fuse_gamma_model\\x94\\x88\\x8c\\x10test_homogeneity\\x94\\x89u\\x8c\\x0eexpected_motif\\x94\\x8c\\x05NNANN\\x94\\x8c\\x0bcomparisons\\x94]\\x94\\x8c\\rfip37_vs_col0\\x94a\\x8c\\tmulticomp\\x94]\\x94\\x8c\\x15fip37_vs_fio1_vs_col0\\x94au\\x8c\\x04rule\\x94\\x8c\\x17generate_yanocomp_plots\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cN/cluster/ggs_lab/mtparker/papers/fiona/fiona_nanopore/rules/notebook_templates\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/cluster/ggs_lab/mtparker/papers/fiona/fiona_nanopore/pipeline');\n",
    "######### snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from glob import glob\n",
    "import random\n",
    "import re\n",
    "import itertools as it\n",
    "from collections import Counter\n",
    "import json\n",
    "import gzip\n",
    "from collections import defaultdict, Counter\n",
    "from snakemake.io import glob_wildcards\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_logo import draw_logo\n",
    "from upsetplot import plot as plot_upset, UpSet\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "from IPython.display import Markdown, display_markdown\n",
    "\n",
    "import pysam\n",
    "import pybedtools as pybt\n",
    "\n",
    "## Default plotting params\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set(font='Arial')\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "style = sns.axes_style('white')\n",
    "style.update(sns.axes_style('ticks'))\n",
    "style['xtick.major.size'] = 2\n",
    "style['ytick.major.size'] = 2\n",
    "sns.set(font_scale=1.2, style=style)\n",
    "pal = sns.color_palette(['#0072b2', '#d55e00', '#009e73', '#f0e442', '#cc79a7', '#56b4e9', '#e69f00'])\n",
    "cmap = ListedColormap(pal.as_hex())\n",
    "sns.set_palette(pal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_formatted_markdown(md, **kwargs):\n",
    "    md_f = md.format(**kwargs)\n",
    "    display_markdown(Markdown(md_f))\n",
    "\n",
    "MD_TEXT = open('../rules/notebook_templates/md_text/upset.md').readlines()\n",
    "\n",
    "display_formatted_markdown(\n",
    "    '# Intersection plots of modification sites detected in Col-0, *fip37-4* and *fio1-1*',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "YANOCOMP_COLUMNS = [\n",
    "    'chrom', 'start', 'end', 'gene_id', 'kmer', 'score', 'strand',\n",
    "    'log_odds', 'log_odds_lower_ci', 'log_odds_upper_ci',\n",
    "    'pval', 'fdr', 'fip37_frac_mod', 'fio1_frac_mod', 'col0_frac_mod',\n",
    "    'g_stat', 'hom_g_stat', 'fip37_vs_fio1_g_stat', 'fip37_vs_col0_g_stat',\n",
    "    'fio1_vs_col0_g_stat', 'fip37_vs_fio1_fdr', 'fip37_vs_col0_fdr',\n",
    "    'fio1_vs_col0_fdr', 'unmod_mean', 'unmod_std', 'mod_mean', 'mod_std',\n",
    "    'ks_stat'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_bed = os.path.join(os.environ['TMPDIR'], 'yanocomp_miclip_dist.bed')\n",
    "\n",
    "pybt.BedTool(snakemake.input.yanocomp_sites[0]).closest(\n",
    "    b=snakemake.input.miclip_peaks,\n",
    "    s=True, D='a'\n",
    ").saveas(tmp_bed)\n",
    "\n",
    "\n",
    "MICLIP_COLUMNS = ['miclip_chrom', 'miclip_start', 'miclip_end', 'miclip_name', 'miclip_score', 'miclip_strand', 'miclip_dist']\n",
    "\n",
    "fip37_fio1_res_miclip_dist = pd.read_csv(\n",
    "    tmp_bed, sep='[\\t:\\[\\],]+',\n",
    "    names=YANOCOMP_COLUMNS + MICLIP_COLUMNS, engine='python'\n",
    ").drop_duplicates(['chrom', 'start', 'end', 'gene_id', 'kmer', 'strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COMPS = ['fip37_vs_fio1', 'fip37_vs_col0', 'fio1_vs_col0']\n",
    "\n",
    "for comp in COMPS:\n",
    "    comp_sig = fip37_fio1_res_miclip_dist.query(f'{comp}_fdr < 0.05').iloc[:, [0, 1, 2, 3, 5, 6]]\n",
    "    comp_sig.to_csv(f'yanocomp/{comp}.posthoc.bed', sep='\\t', header=False, index=False)\n",
    "\n",
    "for comp1, comp2 in it.permutations(COMPS, r=2):\n",
    "    # INTERSECTION\n",
    "    comp_sig = fip37_fio1_res_miclip_dist.query(f'{comp1}_fdr < 0.05 & {comp2}_fdr < 0.05').iloc[:, [0, 1, 2, 3, 5, 6]]\n",
    "    comp_sig.to_csv(f'yanocomp/{comp1}__and__{comp2}.posthoc.bed', sep='\\t', header=False, index=False)\n",
    "    # DIFFERENCE\n",
    "    comp_sig = fip37_fio1_res_miclip_dist.query(f'{comp1}_fdr < 0.05 & {comp2}_fdr > 0.05').iloc[:, [0, 1, 2, 3, 5, 6]]\n",
    "    comp_sig.to_csv(f'yanocomp/{comp1}__not__{comp2}.posthoc.bed', sep='\\t', header=False, index=False)\n",
    "    # DIFFERENCE (with miCLIP support)\n",
    "    comp_sig = fip37_fio1_res_miclip_dist.query(f'{comp1}_fdr < 0.05 & {comp2}_fdr > 0.05 & miclip_dist < 5').iloc[:, [0, 1, 2, 3, 5, 6]]\n",
    "    comp_sig.to_csv(f'yanocomp/{comp1}__not__{comp2}__miclip.posthoc.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fip37_fio1_res_miclip_dist['FIP37_dependent'] = fip37_fio1_res_miclip_dist['fip37_vs_col0_fdr'] < 0.05\n",
    "fip37_fio1_res_miclip_dist['FIO1_dependent'] = fip37_fio1_res_miclip_dist['fio1_vs_col0_fdr'] < 0.05\n",
    "fip37_fio1_res_miclip_dist['miCLIP'] = fip37_fio1_res_miclip_dist['miclip_dist'].abs() < 5\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[0],\n",
    "    n_sites_total=len(fip37_fio1_res_miclip_dist),\n",
    "    perc_fip=fip37_fio1_res_miclip_dist.FIP37_dependent.mean() * 100,\n",
    "    num_fip=fip37_fio1_res_miclip_dist.FIP37_dependent.sum(),\n",
    "    perc_fio=fip37_fio1_res_miclip_dist.FIO1_dependent.mean() * 100,\n",
    "    num_fio=fip37_fio1_res_miclip_dist.FIO1_dependent.sum(),\n",
    ")\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[1],\n",
    "    fip_perc_miclip=fip37_fio1_res_miclip_dist.query('FIP37_dependent').miCLIP.mean() * 100,\n",
    "    fip_num_miclip=fip37_fio1_res_miclip_dist.query('FIP37_dependent').miCLIP.sum(),\n",
    "    num_fio_only=len(fip37_fio1_res_miclip_dist.query('FIO1_dependent & ~FIP37_dependent')),\n",
    "    fio_only_perc_miclip=fip37_fio1_res_miclip_dist.query('FIO1_dependent & ~FIP37_dependent').miCLIP.mean() * 100,\n",
    "    fio_only_num_miclip=fip37_fio1_res_miclip_dist.query('FIO1_dependent & ~FIP37_dependent').miCLIP.sum(),\n",
    ")\n",
    "\n",
    "upset_data = fip37_fio1_res_miclip_dist[['FIP37_dependent', 'FIO1_dependent', 'miCLIP']]\n",
    "upset_data = upset_data.groupby(['FIP37_dependent', 'FIO1_dependent']).apply(\n",
    "    lambda g: pd.Series((len(g), sum(g.miCLIP), g.miCLIP.mean() * 100), index=['total', 'miclip', 'miclip_perc']))\n",
    "upset = plot_upset(\n",
    "    upset_data.total, facecolor='#0072b2', sort_by='degree', sort_categories_by=None, element_size=100,\n",
    "    intersection_plot_elements=4, totals_plot_elements=1,\n",
    ")\n",
    "upset['intersections'].bar(x=np.arange(len(upset_data)), height=upset_data.miclip.values[[0, 2, 1, 3]], width=0.5, color=pal[1], zorder=10)\n",
    "upset['intersections'].legend(['False', 'True'], title='miCLIP support')\n",
    "i = 0\n",
    "for h, _, p in upset_data.iloc[[0, 2, 1, 3]].itertuples(index=False):\n",
    "    upset['intersections'].annotate(s=f'{h:.0f}\\n({p:.1f}%)', xy=(i, h), ha='center', va='bottom')\n",
    "    i += 1\n",
    "\n",
    "upset['intersections'].set_ylabel('Modification sites\\n(intersection size)', rotation=0, ha='right')    \n",
    "\n",
    "totals_data = fip37_fio1_res_miclip_dist[['FIP37_dependent', 'FIO1_dependent', 'miCLIP']].melt(\n",
    "    id_vars=['miCLIP'],\n",
    "    value_vars=['FIP37_dependent', 'FIO1_dependent'],\n",
    "    value_name='total',\n",
    ").query('total').groupby('variable').sum()\n",
    "\n",
    "upset['totals'].barh(y=np.arange(len(totals_data)), width=totals_data.miCLIP.iloc[[1, 0]], height=0.5,\n",
    "                     color=pal[1], zorder=10, align='center')\n",
    "i = 0\n",
    "for m, h in totals_data.iloc[[1, 0]].itertuples(index=False):\n",
    "    p = m / h * 100\n",
    "    upset['totals'].annotate(s=f'{h:.0f}\\n({p:.1f}%)', xy=(h, i), ha='right', va='center')\n",
    "    i += 1\n",
    "upset['totals'].set_xlabel('Modification\\nsites (total)')\n",
    "plt.savefig(snakemake.output.site_upset)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = fip37_fio1_res_miclip_dist.query('FIO1_dependent')\n",
    "m = m.assign(\n",
    "    fip_lod=np.abs(np.log(m.fip37_frac_mod + 1e-3) - np.log(m.col0_frac_mod + 1e-3)),\n",
    "    fio_lod=np.abs(np.log(m.fio1_frac_mod + 1e-3) - np.log(m.col0_frac_mod + 1e-3)),\n",
    ")\n",
    "\n",
    "display_formatted_markdown(\n",
    "    MD_TEXT[2],\n",
    "    fio_perc_fip=fip37_fio1_res_miclip_dist.query('FIO1_dependent').FIP37_dependent.mean() * 100,\n",
    "    fio_num_fip=fip37_fio1_res_miclip_dist.query('FIO1_dependent').FIP37_dependent.sum(),\n",
    ")\n",
    "\n",
    "\n",
    "m = pd.melt(\n",
    "    m, id_vars=[], value_vars=['fip_lod', 'fio_lod'],\n",
    "    var_name='Genotype',\n",
    "    value_name='Effect size (absolute log ratio)'\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "sns.boxplot(\n",
    "    x='Genotype',\n",
    "    y='Effect size (absolute log ratio)',\n",
    "    data=m, palette=[pal[2], pal[1]], showfliers=False\n",
    ")\n",
    "ax.set_xticklabels(['fip37-4', 'fio1-1'])\n",
    "plt.savefig(snakemake.output.site_effect_size)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## U6 m6A IP analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "u6_ip_data = pd.read_csv('ext_data/u6_m6a_ip.tsv', sep='\\t')\n",
    "u6_ip_data = u6_ip_data.pivot_table(index=['genotype', 'bio_rep', 'gene_id'], columns='treatment', values=['ct'], aggfunc=np.nanmean)\n",
    "u6_ip_data.columns = u6_ip_data.columns.droplevel(0)\n",
    "u6_ip_data = u6_ip_data.reset_index()\n",
    "u6_ip_data['dct'] = ((u6_ip_data['input'] - np.log2(2)) - u6_ip_data['IP'])\n",
    "\n",
    "u6_u2_exprs_data = u6_ip_data.pivot_table(index=['genotype', 'bio_rep'], columns='gene_id', values=['input'])\n",
    "u6_u2_exprs_data.columns = u6_u2_exprs_data.columns.droplevel(0)\n",
    "u6_u2_exprs_data = u6_u2_exprs_data.reset_index()\n",
    "u6_u2_exprs_data['dct'] = ((u6_u2_exprs_data['U2']) - u6_u2_exprs_data['U6'])\n",
    "\n",
    "display_formatted_markdown(MD_TEXT[3])\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(12, 5), ncols=2)\n",
    "sns.pointplot(\n",
    "    x='genotype',\n",
    "    y='dct',\n",
    "    data=u6_u2_exprs_data,\n",
    "    order=['col0', 'fio1'],\n",
    "    color='#777777',\n",
    "    join=False,\n",
    "    errwidth=2,\n",
    "    capsize=0.1,\n",
    "    ci='sd',\n",
    "    ax=axes[0]\n",
    "    \n",
    ")\n",
    "sns.stripplot(\n",
    "    x='genotype',\n",
    "    y='dct',\n",
    "    data=u6_u2_exprs_data,\n",
    "    jitter=0.2,\n",
    "    size=8,\n",
    "    order=['col0', 'fio1'],\n",
    "    ax=axes[0]\n",
    ")\n",
    "axes[0].set_xticklabels(['Col-0', 'fio1-1'])\n",
    "axes[0].set_ylabel('Expression relative to U2 (-ΔCt)')\n",
    "axes[0].set_xlabel('Genotype')\n",
    "axes[0].set_title('U6 expression')\n",
    "\n",
    "sns.pointplot(\n",
    "    hue='genotype',\n",
    "    y='dct',\n",
    "    x='gene_id',\n",
    "    data=u6_ip_data,\n",
    "    hue_order=['col0', 'fio1'],\n",
    "    order=['U6', 'U2'],\n",
    "    palette=['#777777', '#777777'],\n",
    "    dodge=0.5,\n",
    "    join=False,\n",
    "    errwidth=2,\n",
    "    capsize=0.1,\n",
    "    ci='sd',\n",
    "    ax=axes[1]\n",
    "    \n",
    ")\n",
    "sns.stripplot(\n",
    "    hue='genotype',\n",
    "    y='dct',\n",
    "    x='gene_id',\n",
    "    data=u6_ip_data,\n",
    "    jitter=0.2,\n",
    "    dodge=0.5,\n",
    "    size=8,\n",
    "    hue_order=['col0', 'fio1'],\n",
    "    order=['U6', 'U2'],\n",
    "    ax=axes[1]\n",
    ")\n",
    "axes[1].set_xticklabels(['U6', 'U2'])\n",
    "axes[1].set_ylabel('Enrichment over input (-ΔCt)')\n",
    "axes[1].set_xlabel('Template')\n",
    "axes[1].set_title('m6A-IP (SySy 202 003)')\n",
    "axes[1].axvline(0.5, ls='--', color='#252525')\n",
    "axes[1].legend_.remove()\n",
    "h1 = axes[1].scatter([], [], color=pal[0], label='Col-0')\n",
    "h2 = axes[1].scatter([], [], color=pal[1], label='fio1-1')\n",
    "axes[1].legend([h1, h2], ['Col-0', 'fio1-1'], loc=4)\n",
    "plt.savefig('figures/u6_m6a_ip_sysy202_003_qpcr.svg')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "u6_ip_data = pd.read_csv('ext_data/u6_m6a_ip_beth.csv')\n",
    "u6_ip_data = u6_ip_data.pivot_table(index=['genotype', 'bio_rep', 'gene_id'], columns='treatment', values=['ct'])\n",
    "u6_ip_data.columns = u6_ip_data.columns.droplevel(0)\n",
    "u6_ip_data = u6_ip_data.reset_index()\n",
    "u6_ip_data['dct'] = ((u6_ip_data['input'] - np.log2(5)) - u6_ip_data['IP'])\n",
    "\n",
    "u6_u2_exprs_data = u6_ip_data.pivot_table(index=['genotype', 'bio_rep'], columns='gene_id', values=['input'])\n",
    "u6_u2_exprs_data.columns = u6_u2_exprs_data.columns.droplevel(0)\n",
    "u6_u2_exprs_data = u6_u2_exprs_data.reset_index()\n",
    "u6_u2_exprs_data['dct'] = ((u6_u2_exprs_data['U2']) - u6_u2_exprs_data['U6'])\n",
    "\n",
    "fig, axes = plt.subplots(figsize=(12, 5), ncols=2)\n",
    "sns.pointplot(\n",
    "    x='genotype',\n",
    "    y='dct',\n",
    "    data=u6_u2_exprs_data,\n",
    "    order=['col0', 'fio1-1', 'fio1-3', 'fio1-4'],\n",
    "    color='#777777',\n",
    "    join=False,\n",
    "    errwidth=2,\n",
    "    capsize=0.1,\n",
    "    ci='sd',\n",
    "    ax=axes[0]\n",
    "    \n",
    ")\n",
    "sns.stripplot(\n",
    "    x='genotype',\n",
    "    y='dct',\n",
    "    data=u6_u2_exprs_data,\n",
    "    jitter=0.2,\n",
    "    size=8,\n",
    "    order=['col0', 'fio1-1', 'fio1-3', 'fio1-4'],\n",
    "    ax=axes[0]\n",
    ")\n",
    "axes[0].set_xticklabels(['Col-0', 'fio1-1', 'fio1-2', 'fio1-3'])\n",
    "axes[0].set_ylabel('Expression relative to U2 (-ΔCt)')\n",
    "axes[0].set_xlabel('Genotype')\n",
    "axes[0].set_title('U6 expression')\n",
    "\n",
    "sns.pointplot(\n",
    "    hue='genotype',\n",
    "    y='dct',\n",
    "    x='gene_id',\n",
    "    data=u6_ip_data,\n",
    "    hue_order=['col0', 'fio1-1', 'fio1-3', 'fio1-4'],\n",
    "    order=['U6', 'U2'],\n",
    "    palette=['#777777', '#777777'],\n",
    "    dodge=0.5,\n",
    "    join=False,\n",
    "    errwidth=2,\n",
    "    capsize=0.1,\n",
    "    ci='sd',\n",
    "    ax=axes[1]\n",
    ")\n",
    "sns.stripplot(\n",
    "    hue='genotype',\n",
    "    y='dct',\n",
    "    x='gene_id',\n",
    "    data=u6_ip_data,\n",
    "    jitter=0.2,\n",
    "    dodge=0.5,\n",
    "    size=8,\n",
    "    hue_order=['col0', 'fio1-1', 'fio1-3', 'fio1-4'],\n",
    "    order=['U6', 'U2'],\n",
    "    ax=axes[1]\n",
    ")\n",
    "axes[1].set_xticklabels(['U6', 'U2'])\n",
    "axes[1].set_ylabel('Enrichment over input (-ΔCt)')\n",
    "axes[1].set_xlabel('Template')\n",
    "axes[1].set_title('m6A-IP (ABE572)')\n",
    "axes[1].axvline(0.5, ls='--', color='#252525')\n",
    "axes[1].legend_.remove()\n",
    "h1 = axes[1].scatter([], [], color=pal[0], label='Col-0')\n",
    "h2 = axes[1].scatter([], [], color=pal[1], label='fio1-1')\n",
    "h3 = axes[1].scatter([], [], color=pal[2], label='fio1-3')\n",
    "h4 = axes[1].scatter([], [], color=pal[3], label='fio1-4')\n",
    "axes[1].legend([h1, h2, h3, h4], ['Col-0', 'fio1-1', 'fio1-3', 'fio1-4'], loc=4)\n",
    "plt.savefig('figures/u6_m6a_ip_ABE572_qpcr.svg')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
